\documentclass[aps,prb,twocolumn,superscriptaddress,floatfix,longbibliography]{revtex4-2}

\usepackage{minted}
\usepackage{amsmath,amssymb} % math symbols
\usepackage{bm} % bold math font
\usepackage{graphicx} % for figures
\usepackage{comment} % allows block comments
%\usepackage{ulem} % allows strikeout text, e.g. \sout{text}

\usepackage{textcomp} % This package gives the text quote '
\usepackage{algpseudocode}
\usepackage{algorithm}
\usepackage{enumitem}
\setlist{noitemsep,leftmargin=*,topsep=0pt,parsep=0pt}

\usepackage{xcolor} % \textcolor{red}{text} will be red for notes
\definecolor{lightgray}{gray}{0.6}
\definecolor{medgray}{gray}{0.4}

\usepackage{hyperref}
\hypersetup{
colorlinks=true,
urlcolor= blue,
citecolor=blue,
linkcolor= blue,
% bookmarks=true,
% bookmarksopen=false,
}
\usepackage{xurl}
\usepackage{float}

% Code to add paragraph numbers and titles
\newif\ifptitle
\newif\ifpnumber
\newcounter{para}
\newcommand\ptitle[1]{\par\refstepcounter{para}
{\ifpnumber{\noindent\textcolor{lightgray}{\textbf{\thepara}}\indent}\fi}
{\ifptitle{\textbf{[{#1}]}}\fi}}
\ptitletrue  % comment this line to hide paragraph titles
\pnumbertrue  % comment this line to hide paragraph numbers

% minimum font size for figurese
\newcommand{\minfont}{6}

% Uncomment this line if you prefer your vectors to appear as bold letters.
% By default they will appear with arrows over them.
% \renewcommand{\vec}[1]{\bm{#1}}

% Allows to rewrite the same title in the supplement
\newcommand{\mytitle}{Efficient rank-based statistic for partially overlapping gene sets}

\begin{document}

\title{\mytitle}

\author{Roc Salvador Andreazini}
\email[]{rocs@stud.ntnu.no}
\affiliation{Department of Computer Science, Norwegian University of Science and Technology, Trondheim, Norway}
\author{Pål Sætrom}
\email[]{pal.satrom@ntnu.no}
\affiliation{Department of Clinical and Molecular Medicine, Norwegian University of Science and Technology, Trondheim, Norway}


\date{\today}

\begin{abstract}
Gene set enrichment analyses use rank-based statistics to test whether related genes are uniformly distributed in an ordered gene list. By testing several gene sets with defined biological functions on gene lists from a biological experiment, one can determine whether specific biological functions are significantly affected by the experiment. The goal of this thesis is to develop efficient algorithms for computing rank-based statistics for thousands of gene sets on tens of thousands of gene lists and use these as a basis for analyzing data from single-cell sequencing experiments.
\end{abstract}

\maketitle
\section{\label{sec:Start}Introduction}

\subsection{Objectives} This thesis aims to develop efficient algorithms for computing rank-based statistics for thousands of gene sets on tens of thousands of gene lists and use these as a basis for analysing data from single-cell sequencing and RNA sequencing experiments.

\subsection{Background}

\ptitle{RNA sequencing} RNA sequencing (RNA-seq) is a powerful experimental technique for studying gene expression and transcriptomics. It allows researchers to quantify and analyse the RNA molecules present in a biological sample, providing insights into the transcriptional activity of genes.

The main analysis goal in RNA-seq is often to identify differentially expressed genes or transcripts between different conditions or experimental groups. Statistical methods, such as the negative binomial-based methods (e.g., DESeq2, edgeR) or maximum likelihood estimation (e.g., Cufflinks), are commonly used to detect significant changes in expression levels. These analyses provide insights into the genes or pathways upregulated or downregulated under specific conditions.

Once differentially expressed genes are identified, further analyses can be performed to understand their biological significance. Gene ontology (GO) enrichment analysis, pathway analysis, and functional annotation can help elucidate the underlying biological processes, molecular functions, and cellular components associated with the differentially expressed genes.


\vspace{2mm}

\ptitle{Single-cell sequencing} \cite{triumphs-and-limitations-scrna} Single-cell sequencing (scRna-seq) examines the sequence information from individual cells with optimized next-generation sequencing technologies, providing a higher resolution of cellular differences and a better understanding of the function of an individual cell in the context of its microenvironment.

These improvements in the technology for sampling expression cells generate some computational problems:
\begin{itemize}
 \item Large datasets: The experiments can capture hundreds of thousands of cells.
 \item Sparse matrices: scRna-seq experiments generate a result matrix with most zero values due to biological and technological limitations.
 \begin{itemize}
    \item Biological heterogeneity: Not all genes are expressed in every cell, resulting in many zero or low expression values.
    \item Dropout events: Dropout refers to the phenomenon where the expression of a gene is not detected or is detected at deficient levels in a single cell, even though the gene is expressed in reality.
 \end{itemize}

\end{itemize}

\subsection{Methods}

\ptitle{Rank-based statistics (ssGSEA)} Single-sample Gene Set Enrichment Analysis (ssGSEA) is a computational method used to assess the activity or enrichment of predefined gene sets within individual samples or observations, such as gene expression profiles. It is an extension of the traditional Gene Set Enrichment Analysis (GSEA) that is typically performed on a cohort or population of samples.

SsGSEA aims to determine the relative activity levels of predefined gene sets within a single sample. This approach is particularly useful when analysing gene expression data from individual patients or experimental conditions, where the focus is on understanding the molecular characteristics of a specific sample rather than comparing it to others.

\vspace{2mm}

For this thesis, the statistic used is a normalized Kolmogorov-Smirnov statistic \cite{diabetes}, and it's defined as follows:

\begin{enumerate}

\begin{figure}[h]
    \includegraphics[clip=true,width=7cm]{img/io-gsea.png}
    \caption{GSEA input and output matrices}
    \label{fig:pixels}
\end{figure}

\item Normalize the expression matrix.
\item Sort the gene counts for each sample in descending order.
\item Compute the enrichment score (\ref{eqn:ES}) for every sample and gene set, being the score for each sampled gene ($S_i$)
\begin{equation}
S_i = \sqrt{\dfrac{N - G} {G}}
\label{eqn:EPS}
\end{equation}

if the gene is in the current gene set

\begin{equation}
S_i = -\sqrt{\dfrac{G} {N - G}}
\label{eqn:ENS}
\end{equation}

if it's not.

\begin{equation}
ES = \max_{{1 \leqslant j < N}}\sum_{i=1}^{N} S_i
\label{eqn:ES}
\end{equation}

Where $N$ is the number of sampled genes, $G$ is the size of the current gene set

\end{enumerate}
\vspace{2mm}

\ptitle{Linear models} \cite{limma} Linear models are used to analyze gene expression data and identify differentially expressed genes between different experimental conditions or groups. The linear model framework provides a flexible and powerful approach to account for various sources of variation and perform statistical inference.

A linear model is defined as:

\begin{equation}
y = X\beta + \epsilon
\end{equation}

where:

\begin{itemize}
\item $y$ represents the observed gene expression data for a particular gene across different samples.
\item $X$ is the design matrix that incorporates information about the experimental conditions or groups. Each column of $X$ represents a factor or covariate associated with the experiment (e.g., treatment, time points).
\item $\beta$  is the vector of coefficients that represents the differential expression effects associated with each factor or condition. Each coefficient $\beta$ corresponds to the effect of a particular factor on gene expression.
\item $\epsilon$ represents the residual error term, accounting for the random variation that cannot be explained by the factors in the model.
\end{itemize}

The goal of the linear model is to estimate the coefficients $\beta$ and test for the significance of the differential expression effects. This is done through the application of statistical methods, such as hypothesis testing or estimation of effect sizes.

\ptitle{Cell clustering} Clustering is a data analysis method that aims to group similar objects while maximizing the dissimilarity between different groups. In single-cell sequencing experiments, clustering is employed to identify distinct cell populations based on their transcriptional profiles. By partitioning the dataset into subsets of cells that exhibit similar gene expression patterns, clustering allows for uncovering hidden structures and relationships within the cellular landscape.

\section{\label{sec:LaTeX} Methods}

\ptitle{gseacc} gseacc is an R package developed for this thesis to run the ssGsea for Rna-seq and scRna-seq datasets.  It uses the Rcpp library to be able to run C++ code in R. The reason why C++ was chosen is because it offers great performance and Objective Oriented programming. As the main features, it supports multithreading, and large data sets can be run in chunks to reduce RAM usage.

The source code and its corresponding documentation can be found in \cite{gseacc-github}.

\subsection{Rna sequencing}

\ptitle{gseacc}

\begin{figure}[H]
\caption{Pseudocode of gsea function for Rna-seq in gseacc}
\begin{algorithmic}[1]
\State{$N$ $\gets$ number of sampled genes}
\State{$GS$ $\gets$ number of gene sets}
\For{$k \gets 1$ to $GS$}
    \State{$G$ $\gets$ $geneSet_k$ size}
    \State{$positive$ $\gets$ $\sqrt{\dfrac{N - G} {G}}$}
    \State{$negative$ $\gets$ -$\sqrt{\dfrac{G} {N - G}}$}
    \For{$j \gets 1$ to $S$}
        \For{$i \gets 1$ to $N$}
            \State{$current$ $\gets$ {$0$}}
            \State{$max$ $\gets$ {$0$}}
            \If{$genes_i$ in $geneSets_k$}
                \State{$current$ $\gets$ $current$ + $positive$}
            \Else
                \State{$current$ $\gets$ $current$ + $negScore$}
            \EndIf
            \State{$max$ $\gets$ $max(max, current)$}
        \EndFor
        \State{$results_{ij}$ $\gets$ max}
    \EndFor
\EndFor
\end{algorithmic}
\end{figure}


\ptitle{Limma} The

\subsection{scRna-seq}

\begin{figure}[H]
\centering
\includegraphics[clip=true,width=4cm]{img/pipeline.png}
\caption{scRna-seq analysis pipeline}
\label{fig:pixels}
\end{figure}

\ptitle{gseacc} The main challenge to running the Enrichment Score with scRna-seq datasets is its large size. This is why gseacc implements a function that allows running the GSEA algorithm in chunks. These chunks are obtained by splitting the expression matrix by samples, in the case of \texttt{Anndata} objects using the \texttt{chunk\_X} function. Since the algorithm is run in different function calls and storing the previous results in the RAM is impossible, different output files are generated. The n most variable gene sets across the samples can be obtained by running the \texttt{filterResults} function from the \texttt{gseacc} package. (\ref{fig:chunkedgsea})


\begin{figure}[h]
\centering
\includegraphics[clip=true,width=7cm]{img/gsea-pipeline.png}
\caption{Chunked ssGsea for scRna-seq}
\label{fig:chunkedgsea}
\end{figure}


\ptitle{Seurat \cite{seurat-v4} \cite{seurat-web}} The main steps in cell clustering using the Seurat package usually are the following:
\begin{itemize}
\item Scale and center data: Scale and center the gene expression matrix to make the gene expression values comparable across cells
\item Principal component analysis (PCA): Use PCA to reduce the dimensionality of the dataset. PCA identifies the principal components that capture the largest sources of variation in the data. These components serve as new features representing the original expression matrix.
\item Find the cell clusters: Construct a nearest-neighbor graph based on the selected principal components, where each cell is connected to its nearest neighbors. Clustering algorithms, such as Louvain or Leiden, are then applied to partition the graph into distinct clusters. These clusters represent putative cell types or states.
\item Reduce dimensionality: Reduce the dimensionality of the clusters using techniques like UMAP (Uniform Manifold Approximation and Projection).
\end{itemize}



\section{\label{sec:Results}Results and analysis}

All R scripts used for each section can be found on GitHub \cite{thesis-github}.

\begin{table}[H]
\centering
\begin{tabular}{ | c | c | }
    \hline
    Processor & Intel® Core™ i5-8250U  \\
    RAM & 8 GB \\
    OS & Ubuntu 23.04 \\
    R version & 4.2 \\
    \hline
\end{tabular}
\caption{Resources used to run all the experiments}
\end{table}


\subsection{RNA-seq}

\ptitle{The dataset \label{pt:rnaseqdata}} For the RNA-seq experiment, the expression matrix from the GSE121212 experiment \cite{gse121212} was used. It contains 147 samples of 60675 genes. This dataset contains gene counts for different samples of human skin and the following phenotypes:

\begin{itemize}
    \item Healthy control skin
    \item Non-lesional psoriasis skin
    \item Lesional psoriasis skin
    \item Non-lesional atopic dermatitis skin
    \item Lesional atopic dermatitis skin
\end{itemize}

\ptitle{Steps} The Enrichment Score matrix was obtained using gseacc \cite{gse121212-go-es} and analysed using Limma to get the most representing gene sets when comparing different phenotypes.

\begin{table}[H]
\centering
\label{tab:pssolvspsol}
\caption{Scalability for the Enrichment Score computation for the GSE121212 experiment and all the 22962 gene sets from the Gene Ontology. The speedup is the improvement in speed of execution respect to the execution with one thread. The overheads in the 8-threads run may be caused due to the fact the processor only has 4 CPUs with 2 threads each.}
\begin{tabular}{ | c @{\hspace{0.6cm}} c @{\hspace{0.5cm}} c | }
    \hline
    Threads & Elapsed time (min) & Speedup \\
    \hline
    \hline
    1 & 398 & 1 \\
    2 & 204 & 1.95 \\
    4 & 111 & 3.59 \\
    8 & 85 & 4.67\\
    \hline
\end{tabular}
\end{table}

\ptitle{Psoriasis non-lesional vs psoriasis lesional} The results for the most differentially expressed gene sets between the samples with Psoriasis
lesional and Psoriasis non-lesional can be found in the table \ref{tab:pssolvspsol}.

The most differentially expressed gene set is GO:2000330. This Gene Ontology term is defined as  any process that activates or increases the frequency, rate or extent of T-helper 17-cell lineage commitment \cite{go:2000330}. This cell type is associated with autoimmune diseases such as multiple sclerosis, rheumatoid arthritis, and psoriasis.

\begin{table}[H]
\centering
\label{tab:pssolvspsol}
\caption{Most different expressed gene sets in Psoriasis non-lesional and Psoriasis-lesional}
\begin{tabular}{ | c @{\hspace{0.6cm}} c @{\hspace{0.5cm}} c @{\hspace{0.5cm}} c | }
    \hline
    GO term & logFC & AveExpr & P value  \\
    \hline
    \hline
    GO:2000330 & -282.9675 & 145.89252 & 1.710876e-33 \\
    GO:0048817 & -267.3064 & 94.67299 &  1.447090e-32 \\
    GO:0008973 & -224.9845 & 89.61911 & 1.647568e-32 \\
    GO:0032500 & -220.1325 & 99.53479 & 3.469039e-32 \\
    GO:0002815 & -220.1325 & 99.53479 & 3.469039e-32 \\
    \hline
    \hline
     & & & \\
     & & & \\
     & & & \\
     & & & \\
     & & & \\
     & & & \\
    \hline
\end{tabular}
\end{table}

\ptitle{Control vs Psoriasis non-lesional}

In the case of CTRL vs Psoriasis non-lesional, the most overrepresented gene is GO:0035373

\begin{table}[H]
\centering
\label{tab:pssolvspsol}
\caption{Most different expressed gene sets in Control and Psoriasis-lesional}
\begin{tabular}{ | c @{\hspace{0.6cm}} c @{\hspace{0.5cm}} c @{\hspace{0.5cm}} c | }
    \hline
    GO term & logFC & AveExpr & P value \\
    \hline
    \hline
    GO:0035373 & 153.1872 & 157.9633 & 3.507739e-08 \\
    GO:1904103 & 100.3478 & 129.0386 & 1.492540e-07 \\
    GO:1904105 & 100.3478 & 129.0386 & 1.492540e-07 \\
    GO:0044595 & -108.4241 & 113.7337 & 5.931994e-07 \\
    GO:0010420 & -108.4241 & 113.7337 &  5.931994e-07 \\
    \hline
    \hline
     & & & \\
     & & & \\
     & & & \\
     & & & \\
     & & & \\
    \hline
\end{tabular}
\end{table}

\subsection{Gene set analysis vs gene analysis}

\begin{figure}[H]
\centering
\includegraphics[clip=true,width=7cm]{img/GSE121212-expr-clusters.png}
\caption{Sample clusters from the GSE121212 expression matrix}
\label{fig:chunkedgsea}
\end{figure}

\begin{figure}[H]
\centering
\includegraphics[clip=true,width=7cm]{img/GSE121212-expr-heatmap.png}
\caption{Sample clusters from the GSE121212 expression matrix}
\label{fig:chunkedgsea}
\end{figure}

\begin{figure}[H]
\centering
\includegraphics[clip=true,width=7cm]{img/GSE121212-ES-clusters.png}
\caption{Sample clusters from the GSE121212 Enrichment Score matrix}
\label{fig:chunkedgsea}
\end{figure}

\begin{figure}[H]
\centering
\includegraphics[clip=true,width=7cm]{img/GSE121212-ES-heatmap.png}
\caption{Sample clusters from the GSE121212 Enrichment Score matrix}
\label{fig:chunkedgsea}
\end{figure}
\subsection{scRNA-seq}

\ptitle{The dataset} The dataset used to test the scRna-seq GSEA implementation consists of $1.9\cdot10^5$ samples (cells) and $2\cdot10^4$ genes from adult human skin. It was extracted from \textit{Developmental cell programs are co-opted in inflammatory skin disease}\cite{healthy}.

It has approximately $10\cdot10^9$ counts, that is, $10^3$ orders of magnitude larger than the Rna-seq dataset \ref{pt:rnaseqdata}.

\ptitle{Steps} The first step, as in the Rna-seq experiment, was to run the GSEA. The running time was 9h and 46min using 8 threads, with an average of 4.9 GB of RAM usage, with the expression matrix split into 100 chunks. Then the Enrichment Score results were filtered by choosing the most variable gene sets. This first section was run using \texttt{gseacc}.


\begin{figure}[H]
\centering
\includegraphics[clip=true,width=6cm]{img/gene-sets-variance.png}
\caption{Enrichment Score variance of the gene sets of the Gene Ontology. There are ~1000 gene sets with a high variance, ~9000 with low variance and ~10000 with no variance at all.}
\label{fig:pixels}
\end{figure}

\ptitle{Cell clustering}

\section{Conclusions}

\subsection{Limitations}

\bibliography{report}

\end{document}
